source/h2_connection.c

```diff
@@ -256,16 +272,326 @@ static void s_handler_destroy(struct aws_channel_handler *handler) {
+    msg->on_completion = s_on_channel_write_complete;
+    msg->user_data = connection;
+
+    CONNECTION_LOGF(
+        TRACE,
+        connection,
+        "Outgoing frames task acquired message with %zu bytes available",
+        msg->message_data.capacity - msg->message_data.len);
+
+    /* Track number of frames encoded, just used for logging */
+    size_t num_frames_encoded = 0;
+
+    /* Write as many frames from outgoing_frames_queue as possible. */
+    while (!aws_linked_list_empty(outgoing_frames_queue)) {
+        struct aws_linked_list_node *frame_node = aws_linked_list_front(outgoing_frames_queue);
+        struct aws_h2_frame_base *frame = AWS_CONTAINER_OF(frame_node, struct aws_h2_frame_base, node);
+
+        /* #TODO actual functionality to query min required space for a frame */
+        const size_t min_required_bytes = 1024;
+        const size_t available_bytes = msg->message_data.capacity - msg->message_data.len;
+        if (available_bytes < min_required_bytes) {
+            if (msg->message_data.len == 0) {
+                /* We're in trouble if an empty message isn't big enough for this frame to do any work with */
+                CONNECTION_LOGF(
+                    ERROR,
+                    connection,
+                    "Cannot encode %s frame requiring %zu bytes, max available space is %zu",
+                    aws_h2_frame_type_to_str(frame->type),
+                    min_required_bytes,
+                    available_bytes);
+                aws_raise_error(AWS_ERROR_INVALID_STATE);
+                goto error;
+            }
+
+            CONNECTION_LOG(TRACE, connection, "Outgoing frames task filled message, and has more frames to send later");
+            goto done_encoding;
+        }
+
+        /* #TODO some way for frame to say it's not done yet.
+         * Necessary for HEADERS that will split across CONTINUATION frames */
+        if (aws_h2_encode_frame(&connection->thread_data.encoder, frame, &msg->message_data)) {
+            CONNECTION_LOGF(
+                ERROR,
+                connection,
+                "Error encoding frame of type %s: %s",
+                aws_h2_frame_type_to_str(frame->type),
+                aws_error_name(aws_last_error()));
+            goto error;
+        }
+
+        /* Done encoding frame, pop from queue and cleanup*/
+        aws_linked_list_remove(frame_node);
+        aws_h2_frame_clean_up(frame);
+        aws_mem_release(connection->base.alloc, frame);
+
+        num_frames_encoded++;
+    }
+
+    /* Write as many DATA frames from outgoing_streams_list as possible.
+     * We simply round-robin through available streams, instead of using stream priority.
+     *
+     * Respecting priority is not required (RFC-7540 5.3), so we're ignoring it for now. This also keeps use safe
+     * from priority DOS attacks: https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2019-9513
+     */
+    while (!aws_linked_list_empty(outgoing_streams_list)) {
```

Comment:
and msg->msg_data.capacity - msg->msg_data.len > 0?

Reply:
`outgoing_streams_list` is only for streams that *currently* have data to send (there's another datastructure `active_streams_map` that contains ALL living streams). In a future PR there will be some trigger somewhere that pops streams out of this list when they're done sending data.